### ISSUE TITLE
Infinite recursion when using `GraphQL::Query#result` in an `after_query` hook

### ISSUE THREAD

**Issue by jonleighton** [rmosolgo/graphql-ruby#1747]
Hello!

GraphQL version: 1.8.6

I am using `GraphQL::Query#result` in an `after_query` hook, and this can cause infinite recursion. The code looks something like this:

``` ruby
class MySchema < GraphQL::Schema
    instrument(:query, QueryLogger)
end

class QueryLogger
  def self.after_query(query)
    Rails.logger.debug(query.result) if Rails.env.development?
  end
end
```

`GraphQL::Query#result` tries to execute the query if it has not already executed:

https://github.com/rmosolgo/graphql-ruby/blob/be5439d76b4ee2176253eaab9fd329ebd09f5a7a/lib/graphql/query.rb#L166-L170

I had a case where an error in my schema definition prevented the query from executing. However, the `after_query` hook still ran. But because `@executed` is false, `GraphQL::Query#result` tries to execute the query again, which then calls the after hook, and so on...

Note that this problem only seems to happen when the error is related to the *definition* of the schema, rather than the execution of an already well-defined schema.

**Comment by jonleighton** [comment]
Perhaps by having an `@execution_started` boolean?

**Comment by rmosolgo** [comment]
Yes, sounds reasonable to me!

**Comment by rmosolgo** [comment]
I've been working on a big change to how GraphQL schemas are initialized. In 1.10, class-based schemas are built _at load time_. In previous arrangement, some parts of initialization were deferred (for example, caching a list of all types in the schema), and that resulted in errors coming at odd times, like described here. I've added a few more tests around _query_ errors in 65f8027a8, but with this new approach to schema definition available in 1.10, I don't plan to work on this anymore.
